controller/src/scheduler.rs

```diff
@@ -0,0 +1,107 @@
+use cron::Schedule;
+use snafu::ResultExt;
+use std::env;
+use std::str::FromStr;
+
+// Defines the cron expression scheduler related env variable names
+const SCHEDULER_CRON_SECONDS_ENV_VAR: &str = "SCHEDULER_CRON_SECONDS";
+const SCHEDULER_CRON_MINUTES_ENV_VAR: &str = "SCHEDULER_CRON_MINUTES";
+const SCHEDULER_CRON_HOURS_ENV_VAR: &str = "SCHEDULER_CRON_HOURS";
+const SCHEDULER_CRON_DAY_OF_MONTH_ENV_VAR: &str = "SCHEDULER_CRON_DAY_OF_MONTH";
+const SCHEDULER_CRON_MONTH_ENV_VAR: &str = "SCHEDULER_CRON_MONTH";
+const SCHEDULER_CRON_DAY_OF_WEEK_ENV_VAR: &str = "SCHEDULER_CRON_DAY_OF_WEEK";
+const SCHEDULER_CRON_YEARS_ENV_VAR: &str = "SCHEDULER_CRON_YEARS";
+
+/// The module-wide result type.
+type Result<T> = std::result::Result<T, scheduler_error::Error>;
+
+struct BrupopCronScheduler {
+    seconds: String,
+    minutes: String,
+    hours: String,
+    day_of_month: String,
+    month: String,
+    day_of_week: String,
+    year: String,
+}
+
+impl BrupopCronScheduler {
+    pub fn new() -> Self {
+        let scheduler_cron_seconds = read_env_var(SCHEDULER_CRON_SECONDS_ENV_VAR).unwrap();
+        let scheduler_cron_minutes = read_env_var(SCHEDULER_CRON_MINUTES_ENV_VAR).unwrap();
+        let scheduler_cron_hours = read_env_var(SCHEDULER_CRON_HOURS_ENV_VAR).unwrap();
+        let scheduler_cron_day_of_month =
+            read_env_var(SCHEDULER_CRON_DAY_OF_MONTH_ENV_VAR).unwrap();
+        let scheduler_cron_month = read_env_var(SCHEDULER_CRON_MONTH_ENV_VAR).unwrap();
+        let scheduler_cron_day_of_week = read_env_var(SCHEDULER_CRON_DAY_OF_WEEK_ENV_VAR).unwrap();
+        let scheduler_cron_years = read_env_var(SCHEDULER_CRON_YEARS_ENV_VAR).unwrap();
+
+        BrupopCronScheduler {
+            seconds: scheduler_cron_seconds,
+            minutes: scheduler_cron_minutes,
+            hours: scheduler_cron_hours,
+            day_of_month: scheduler_cron_day_of_month,
+            month: scheduler_cron_month,
+            day_of_week: scheduler_cron_day_of_week,
+            year: scheduler_cron_years,
+        }
+    }
+
+    // generate a valid cron expression format string value based on provided env variables.
+    pub fn generate_cron_expression(&self) -> String {
```

Comment:
Can we do anything here (or during (`new()`)) to validate that the cron expression is indeed valid? Seems like there may not be controls in place to protect against it.

Not sure how feasible it is, but maybe we could use the `#[validate()]` macro in the model to assert that each field is valid?  (see the `BottlerocketShadowSpec` for an example). That should protect customers from invalid environment variables very early on. Even with that, it would be a good idea to protect against that case in the code too.

also, nit, use docstring comments, especially on public functions:
```suggestion
    /// Generate a valid cron expression format string value based on provided env variables.
    pub fn generate_cron_expression(&self) -> String {
```

Reply:
>Can we do anything here (or during (new())) to validate that the cron expression is indeed valid?

I discuss the cron expression validation with @jpmcb offline, and we think it might be worthy to validate it when users generate brupop resource yamls. And he is working on helm chart so I'll skip this step here. when the helm chart comes out, I'll re-think how we validate cron expression over there. 

>Not sure how feasible it is, but maybe we could use the #[validate()] macro in the model to assert that each field is valid?

That's a good suggestion. However, for each cron expression elements (minutes, hours....), they have different, wide, and varied usages. The regex will be complicated so that I don't think we should use regex to validate that inputs.

Reply:
I didn't handle validation of cron expression since we will rely on `Schedule::from_str()`.